### ISSUE TITLE
Support multiple projects' dbt docs

### ISSUE THREAD

**Issue by dwreeves** [astronomer/astronomer-cosmos#862]
Discussion here: https://github.com/astronomer/astronomer-cosmos/pull/737#issuecomment-1964349542

Better to get this done now before 1.4 ships than after, I think.

tldr: The main worry I have is that the API for supporting multiple projects' docs can get very convoluted. Ini file format / env vars for config do not natively support arrays of objects, so you need to do something that isn't pretty or is complicated (e.g. read from a yaml file) basically no matter what.

I'm leaving this as an open ended issue to discuss solutions.

**Comment by dwreeves** [comment]
So one implementation that is fully compatible with the existing API would be to do something like this:

- User can provide comma separated list of paths.
- If they do, then create list of each dbt project and that is the landing page for the docs.
- If they don't, then render the docs normally.
- If there is one or none connection_id, then apply to all items in the list.
- If there are multiple comma separated connections, then `zip()` with the paths.

The main challenge here is getting the names of the dbt projects. This requires downloading each `json` and then retrieving the `manifest["metadata"]["project_name"]`. It could be prudent, albeit minor over-optimizing, to use `flask-caching` (and allow the user to override the default time for the cache to 0 if they don't want this) because this does end up downloading a lot of files. (Note: `flask-caching` is already a dependency of Airflow.) Another option is to allow users to pass in their own labels.

If reading from `manifest.metadata.project_name` to get the project names for each doc, you will want to have some sort of exception handling or timeout logic perhaps if there is an issue, if say S3 docs aren't loading but the local ones are. I dunno. Maybe that is too complicated.

The issue with creating a new menu item for dbt docs is twofold. First, it's not appropriate for most users with just one project. Two, dbt project names cannot realistically be automated in this context as loading the `manifest.json` can block the Airflow UI from loading on normal, non-dbt docs pages, and also wastes S3 reads. Or worse: imagine a scenario where your `manifest.json` is messed up, and your entire Airflow UI crashes because the plugin is attempting to read a corrupt or nonexistent JSON, but you also need access to the Airflow UI to diagnose the problem... not good. You cannot automate the names of the UI elements from the `manifest.json`s _if_ you are doing an app-builder menu item approach.

I'm not a happy camper setting any of this up because `airflow.cfg` (`.ini` file format) isn't well suited for this (`.ini` doesn't natively have an array type, which is the proper data model for this), and also it feels like there is no way to avoid that one of the two setups (solo docs, or multi docs) ends up with the suboptimal end of the stick. most users just have one dbt project, so it makes sense to have an abstraction that prioritizes single project deploys (which the current API does well) and let people with complicated setups deal with a slightly complicated API, which is fair and even congruent with what they're doing already. I'm thinking this is the least intrusive way to support multiple projects while keeping the user experience friendly to the majority of users with just one project. (Also, some users are on multiple projects because large projects execute slowly, which is also being addressed in 1.4.)

**Comment by tatiana** [comment]
@dwreeves, we aim to release Cosmos 1.6 by the end of the month. Do you think you may have the bandwidth to work on this before? If not, what do you think about moving it to the 1.7 release or after?

**Comment by benjy44** [comment]
I would be interested by this feature, any progress on it?

**Comment by pankajkoti** [comment]
While discussing [issue #1587](https://github.com/astronomer/astronomer-cosmos/issues/1587) with @bbovenzi and @pierrejeambrun—specifically how to support dbt Docs in Airflow 3 using the proposed external_views and iframe functionality — Upon @tatiana's suggestion, I also followed-up with this broader question of whether Airflow 3 could natively support use cases like rendering separate sub-menu items for multiple dbt projects or something else. Unfortunately, this doesn’t seem likely in the near term.

@pierrejeambrun mentioned the below:

"_Both in AF2 and AF3 plugins are collected a the start of the api server via the PluginManager. Those plugins are then cached and reused through the entire life of the process.
This means that there is no way to modify plugins at runtime (and I don't think we should). One possibility though is to do this collection and addition at the start -> writing a custom plugin that will fetch external resources (fetch the list from somewhere) and then populate the external_views correctly. This will slightly delay the initialisation of the server but you will have all of the views for the different dbt docs projects. This will be achievable just on the Plugin authoring side._"

Following further discussions with @pierrejeambrun and @tatiana, and also gathering ideas discussed in the above issue trail, I am summarising a few ideas to make this work:

Cosmos could collect the list of dbt projects, their dbt docs folder and connection to access the docs folder through:
1. Airflow Variables (this could be tricky as Variables may not exist yet when plugins are getting registered and it would necessitate restart of the server once the Variable is defined)
2. Environment Variables
3. Via a JSON/YAML file containing a mapping of:
	- dbt project name
	- Path to the docs folder
	- Airflow connection ID to access the docs
	- (Optional) Menu item name - explained the need for this in alternative 2 below for Airflow UI interface.

        This file could either be bundled in the deployment image or hosted remotely, with environment variables or settings pointing to it.

In terms of the Airflow UI, two interface options were discussed:
1. A single landing page (Airflow 3 external_view) that links to iframe views for each dbt project. (Personally, I’m not in favour of this — it feels non-native and may result in a poor UX.)
2. Dynamically creating individual menu items for each dbt project based on the gathered mapping by authoring the plugin to support that. To me, this seems like the cleaner and more intuitive approach. We could also collect the menu item name from users in the above mappings, the default could be "dbt docs (<dbt_project_name>)"

Open to more ideas and brainstorming here, would love the community's feedback on this.

**Comment by pankajkoti** [comment]
cc: @dwreeves Please let us know your thoughts on the above 👆🏽 and in case if you'd like to discuss this further with the team would be great to make progress on this feature

**Comment by bbovenzi** [comment]
I like the MenuItem idea. Especially because the UI will consolidate plugins into a menu if there are too many of them to fit in the nav bar. So it can be "dbt docs" -> "project_1", "project_2", etc

**Comment by pankajkoti** [comment]
@bbovenzi are we planning to support the  "dbt docs" -> "project_1", "project_2", etc feature anytime soon with Airflow UI?

**Comment by pierrejeambrun** [comment]
Something like this, with multiple dbt project grouped under a DBT docs tab and embedding an iframe is already possible on latest main of the OSS repo. You can play with it, this is done via `external_views` attribute in plugins. (Theres some documentation and examples too on latest main on how to achieve that).

This will be released in 3.1.0.

![Image](https://github.com/user-attachments/assets/8bd26014-2b30-4cf8-b7d0-d8516520f2b2)

**Comment by bbovenzi** [comment]
Yes, everything with iframes is now in main. You can also add keywords like {DAG_ID}, {TASK_ID}, {RUN_ID} and {MAP_INDEX} to your iframe url and airflow will fill them, Although we don't have the iframe push any state to the airflow url.

**Comment by pankajkoti** [comment]
Support for multiple dbt projects has been added using the "external views" feature of Airflow 3.1 plugins in PR #2009 (planned to be released in Cosmos 1.11.0 release)

Thanks @pierrejeambrun and @bbovenzi for the amazing feature support added in Airflow 3.
